File: leetcode/exercises/1.4.1.medium.group_anagrams_2.py
# -----------------------
# Created       : 20/10/2025
# Last Edited   : 24/10/2025
# Big O         :
# Topics        : 
# Problem Id    : 49
# -----------------------

from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs: list[str]) -> list[list[str]]:
        sets = defaultdict(list)
        for i in strs:

            count = [0] * 26
            for j in i:
                count[ord(j) - ord("a")] += 1

            sets[tuple(count)].append(i)

        return list(sets.values())

sol = Solution()
print(sol.groupAnagrams(["eat","tea","tan","ate","nat","bat"]))

File: leetcode/exercises/1.4.1.medium.group_anagrams_2.go
// -----------------------
// Created     : 02/11/2025
// Last Edited : 02/11/2025
// Topics      : 
// Big O       :
// Problem Id  : 49
// References  :
// Notes       :
// -----------------------

package main

import (
    "fmt"
    "maps"
    "slices"
)

func groupAnagrams(strs []string) [][]string {

    m := make(map[[26]int][]string) 

    for _, v := range strs {

        key := [26]int{}
        for i := 0; i < len(v); i++ {
            key[v[i] - 'a'] = key[v[i] - 'a'] + 1
        }
        m[key] = append(m[key], v)
    }

    vals := maps.Values(m)

    return slices.Collect(vals)
}

func main() {
    fmt.Println(groupAnagrams([]string{"eat","tea","tan","ate","nat","bat"}))
}
